src/session/session_api.c

```diff
@@ -1583,55 +1584,94 @@ __wt_session_range_truncate(
      * object's name space without knowing exactly what records currently appear in the object. For
-     * this reason, do a search-near, rather than a search. Additionally, we have to correct after
-     * calling search-near, to position the start/stop cursors on the next record greater than/less
-     * than the original key. If we fail to find a key in a search-near, there are no keys in the
-     * table. If we fail to move forward or backward in a range, there are no keys in the range. In
-     * either of those cases, we're done.
+     * this reason, position the start/stop cursors to the actual first/last keys in the truncation
+     * range. If possible, we implement this using bounded cursors to prevent false conflicts with
+     * transactions that touch keys right outside of the truncation range. If the table does not
+     * support bounded cursors, we have to fall back to using search-near, but that's not preferable
+     * as it would access the first key following the range, which could result in a false conflict.
      *
      * No need to search the record again if it is already pointing to the btree.
      */
-    if (start != NULL && !F_ISSET(start, WT_CURSTD_KEY_INT))
-        if ((ret = start->search_near(start, &cmp)) != 0 ||
+    if (!F_ISSET(start, WT_CURSTD_KEY_INT)) {
+        if (supports_bounds) {
+            if (!local_start) {
+                /*
+                 * Use a new cursor, because at this level, we can't set bounds on a positioned
+                 * cursor, and at this abstraction level, we can't use WT_CURSOR_IS_POSITIONED to
+                 * fully check.
+                 */
+                WT_ERR(
+                  __session_open_cursor((WT_SESSION *)session, actual_uri, NULL, NULL, &start));
+                trunc_info->start = start;
+                local_start = true;
+            }
+            if (orig_start_key != NULL) {
+                __wt_cursor_set_raw_key(start, orig_start_key);
+                WT_ERR(start->bound(start, "bound=lower"));
+            }
+            if (orig_stop_key != NULL) {
+                __wt_cursor_set_raw_key(start, orig_stop_key);
+                WT_ERR(start->bound(start, "bound=upper"));
+            }
+            WT_ERR_NOTFOUND_OK(start->next(start), true);
+            if (ret == WT_NOTFOUND) {
+                /* If there are no elements, there is nothing to do. */
+                ret = 0;
+                log_trunc = true;
+                goto done;
+            }
+        } else if (orig_start_key == NULL) {
+            WT_ERR_NOTFOUND_OK(start->next(start), true);
+            if (ret == WT_NOTFOUND) {
+                ret = 0;
+                log_trunc = true;
+                goto done;
+            }
+        } else if ((ret = start->search_near(start, &cmp)) != 0 ||
           (cmp < 0 && (ret = start->next(start)) != 0)) {
             WT_ERR_NOTFOUND_OK(ret, false);
```

Comment:
There are 3 distinct places where we walk the cursor next and then do these same operations following it. I tried to think of a way to de-duplicate the code but got sidetracked by baron duties. Can you think of anything to improve this?

Reply:
Would this be an improvement? (We can avoid using the new boolean variable, if we're willing to use something like `goto skip_next`.)
```
    if (!F_ISSET(start, WT_CURSTD_KEY_INT)) {
        needs_next_prev = true;
        if (supports_bounds) {
            if (!local_start) {
                /*
                 * Use a new cursor, because at this level, we can't set bounds on a positioned
                 * cursor, and at this abstraction level, we can't use WT_CURSOR_IS_POSITIONED to
                 * fully check.
                 */
                WT_ERR(
                  __session_open_cursor((WT_SESSION *)session, actual_uri, NULL, NULL, &start));
                trunc_info->start = start;
                local_start = true;
            }
            if (orig_start_key != NULL) {
                __wt_cursor_set_raw_key(start, orig_start_key);
                WT_ERR(start->bound(start, "bound=lower"));
            }
            if (orig_stop_key != NULL) {
                __wt_cursor_set_raw_key(start, orig_stop_key);
                WT_ERR(start->bound(start, "bound=upper"));
            }
        } else if (orig_start_key != NULL) {
            WT_ERR_NOTFOUND_OK(start->search_near(start, &cmp), true);
            if (ret == WT_NOTFOUND) {
                ret = 0;
                log_trunc = true;
                goto done;
            }
            needs_next_prev = (cmp < 0);
        }
        if (needs_next_prev) {
            WT_ERR_NOTFOUND_OK(start->next(start), true);
            if (ret == WT_NOTFOUND) {
                /* If there are no elements, there is nothing to do. */
                ret = 0;
                log_trunc = true;
                goto done;
            }
        }
    }
```

Reply:
It's better albeit unfortunate that the exact same logic exists above for the first search_near call. I considered a macro like `WT_ERR_NOTFOUND_EXIT` or something else but its hard to work in.

Please add in what you have.